src/main/java/com/atlassian/jira/cloud/jenkins/deploymentinfo/service/ChangeLogIssueKeyExtractor.java

```diff
@@ -33,12 +33,8 @@ public Set<String> extractIssueKeys(
         final List<ChangeLogSet<? extends ChangeLogSet.Entry>> changeSets =
                 new ArrayList<>(workflowRun.getChangeSets());
 
-        // Go through all previously failed builds and collect their changesets.
-        WorkflowRun previous = workflowRun.getPreviousBuild();
-        while (Objects.nonNull(previous) && !isBuildSuccessful(previous)) {
-            changeSets.addAll(previous.getChangeSets());
-            previous = previous.getPreviousBuild();
-        }
+        addFirstNonEmptyChangeSetFromPreviousBuilds(workflowRun, changeSets);
+        addChangeSetsFromFailedPreviousBuilds(workflowRun, changeSets);
```

Comment:
Not sure what `changeSet` is here, but what if a workflowRun passes both conditions? 
i.e. doesn't have `changeSets` and doesn't have any previous `isBuildSuccessful`. 
Can this happen? 
Cause if it is possible, then that entry will be added twice. 
Isn't it better to just append this `changeSet` condition with an `OR` instead of doing the while loop twice?

Reply:
mmmmm kinda, 
Changesets is Jenkins copy of whats different in git since the last build.
So if i commit a change, then run a build 10 times, only the first build will contain a changeSet

Ive flipped the order so the new check happens(no changesets) after the failed build call.
Therefore we will never get a double up, if something gets added in the failed build step the subsequent function will get skipped.

Then even it doesn't catch anything thats good we want to keep looping.

Going to type it out to make sure lol
[F] - failed build
[S] - failed build
[C] - failed build
0 - no change set
1 - has changeset


[S1][F1][F0][F1][F0][F0][F0][C]
Will behave same as original code

[S1][S0][S0][S0][S0][S0][C]

Previously failing to get Issuekeys, will loop to the S1